File: Linked_List_free_code_camp/6.linked_list_reverse.cpp
#include<bits/stdc++.h>
using namespace std;

class Node {
public:
    int val;
    Node* next;

    Node(int val, Node* next = NULL) {
        this->val = val;
        this->next = next;
    }
};

void printLinkedListIterative(Node* head) {
    Node* t = head;
    while (t) {
        cout << t->val << " ";
        t = t->next;
    }
}

Node* reverseIterative(Node* head) {
    Node* prev = NULL;
    Node* cur = head;
    Node* next = head->next;

    while (cur) {
        next = cur->next;
        cur->next = prev;
        prev = cur;
        cur = next;
    }

    return prev;
};

Node* reverseRecursive(Node* curr, Node* prev = NULL) {
    if (curr == NULL) return prev;

    Node* newHead = reverseRecursive(curr->next, curr); // faith

    curr->next = prev;

    return newHead;
};


int main() {

    Node* head =
        new Node(10,
            new Node(20,
                new Node(30,
                    new Node(40))));
    // head = new Node(10);

    // head = reverseIterative(head);
    // printLinkedListIterative(head);

    head = reverseRecursive(head);
    printLinkedListIterative(head);


    return 0;
}

File: Linked_List_free_code_camp/6.linked_list_reverse.js
class Node {
    constructor(val, next = null) {
        this.val = val;
        this.next = next;
    }
}

const head =
    new Node(1,
        new Node(2,
            new Node(3,
                new Node(4,
                    new Node(5, null)))));

// A -> B -> C -> D -> null

const printLinkedListIterative = (head) => {
    let curr = head;
    while (curr) {
        console.log(curr.val);
        curr = curr.next;
    }
};

const reverseList = (head) => {
    let prev = null;
    let curr = head;
    let next = head.next;

    while (curr) {
        next = curr.next;
        curr.next = prev;
        prev = curr;
        curr = next;
    }

    return prev;
};

const reverseListRecursive = (curr, prev = null) => {
    if (curr == null) return prev;

    let next = curr.next;
    curr.next = prev;
    let newHead = curr;

    newHead = reverseListRecursive(next, curr);

    return newHead;
};

// printLinkedListIterative(reverseList(head));
printLinkedListIterative(reverseListRecursive(head));